lib/chef/key.rb

```diff
@@ -0,0 +1,203 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require 'chef/json_compat'
+require 'chef/mixin/params_validate'
+
+class Chef
+  class Key
+
+    include Chef::Mixin::ParamsValidate
+
+    attr_reader :actor_field_name
+
+    def initialize(actor, actor_field_name)
+      # Actor that the key is for, either a client or a user.
+      @actor = actor
+
+      unless actor_field_name == "user" || actor_field_name == "client"
+        raise ArgumentError.new("the second argument to initialize must be either 'user' or 'client'")
+      end
+
+      @actor_field_name = actor_field_name
+
+      @name = nil
+      @public_key = nil
+      @expiration_date = nil
+    end
+
+    def chef_rest
+      @rest ||= if @actor_field_name == "user"
+                  Chef::REST.new(Chef::Config[:chef_server_root])
+                else
+                  Chef::REST.new(Chef::Config[:chef_server_url])
+                end
+    end
+
+    def api_base
+      @api_base ||= if @actor_field_name == "user"
+                            "users"
+                          else
+                            "clients"
+                          end
+    end
+
+    def actor(arg=nil)
+      set_or_return(:actor, arg,
+                    :regex => /^[a-z0-9\-_]+$/)
+    end
+
+    def name(arg=nil)
+      set_or_return(:name, arg,
+                    :regex => /^[a-z0-9\-_]+$/)
+    end
+
+    def public_key(arg=nil)
+      set_or_return(:public_key, arg,
+                    :kind_of => String)
+    end
+
+    def expiration_date(arg=nil)
+      set_or_return(:expiration_date, arg,
+                    :regex => /^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z|infinity)$/)
+    end
+
+    def to_hash
+      result = {
+        @actor_field_name => @actor,
+      }
+      result["name"] = @name if @name
+      result["public_key"] = @public_key if @public_key
+      result["expiration_date"] = @expiration_date if @expiration_date
+      result
+    end
+
+    def to_json(*a)
+      Chef::JSONCompat.to_json(to_hash, *a)
+    end
+
+    def create
+      if @public_key.nil? || @expiration_date.nil?
+        raise ArgumentError.new("public_key and expiration_date fields must be populated when create is called")
+      end
+
+      # defaults the key name to the fingerprint of the key
+      if @name.nil?
+        @name = generate_fingerprint(@public_key)
+      end
+
+      payload = {"name" => @name, "public_key" => @public_key, "expiration_date" => @expiration_date}
+      new_key = chef_rest.post_rest("#{api_base}/#{@actor}/keys", payload)
+      Chef::Key.from_hash(new_key)
+    end
+
+    def generate_fingerprint(public_key)
```

Comment:
Sorry to nitpick, but I think in general we prefer for this method to not take an argument, be named `fingerprint` and just operate on the instance variable.